cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java

```diff
@@ -68,25 +67,28 @@
   CloseableClassLoader createClassLoader(Location artifactLocation) throws IOException {
     final File unpackDir = BundleJarUtil.unJar(artifactLocation, DirUtils.createTempDir(tmpDir));
 
-    ClassLoader parentClassLoader;
-
-    ProgramRuntimeProvider sparkRuntimeProvider = runtimeProviderLoader.get(ProgramType.SPARK);
-    if (sparkRuntimeProvider != null) {
-      // Always have spark classes visible for artifact class loading purpose since we don't know if
-      // any classes inside the artifact is a Spark program
-      parentClassLoader = new FilterClassLoader(sparkRuntimeProvider.getClass().getClassLoader(),
-                                                sparkRuntimeProvider.createProgramClassLoaderFilter(ProgramType.SPARK));
-    } else {
-      // If no Spark is available, just use the standard filter
-      parentClassLoader = new FilterClassLoader(getClass().getClassLoader(), FilterClassLoader.defaultFilter());
+    ProgramClassLoader programClassLoader = null;
+    try {
+      // Try to create a ProgramClassLoader from the Spark runtime system if it is available.
+      // It is needed because we don't know what program types that an artifact might have.
+      ProgramRunner programRunner = programRunnerFactory.create(ProgramType.SPARK);
+      if (programRunner instanceof ProgramClassLoaderProvider) {
+        programClassLoader = ((ProgramClassLoaderProvider) programRunner).createProgramClassLoader(cConf, unpackDir);
+      }
+    } catch (IllegalArgumentException e) {
+      // If Spark is not supported, exception is expected. We'll use the default filter.
```

Comment:
Since ``IllegalArgumentException`` is pretty generic, should we only enclose line 74 in this ``try...catch``? Otherwise throw a specific exception from ``DefaultProgramRunnerFactory.create(ProgramType)``?